### PR TITLE
Add Core Ruleset modules for NBI

### PR DESCRIPTION
**Pull request by CCDevelopForFun**
## Context
Implements a persistent AI behavior guidance system through hierarchical rules that maintain context across sessions and ensure consistent responses. Rules are organized globally and by mode (ask/agent/inline-chat) with smart matching based on file patterns, kernels, and cell types.

## Changes
- **Core models**: `Rule`, `RuleSet`, `NotebookContext`, `RuleScope` data structures
- **Rule management**: Discovery, loading, validation, and application logic
- **Config integration**: Extended NBIConfig with ruleset properties
- **Directory structure**: `~/.jupyter/nbi/rules/` (Adjustable) with global and mode-specific rules

## Test
- 49 comprehensive tests covering all functionality
- Error handling and edge cases validated
- Full integration with existing config system
- Performance optimized pattern matching

Ready for production deployment.

### CODE REVIEW

File: notebook_intelligence/config.py [hunk 1]
```diff
@@ -120,3 +120,38 @@ def store_github_access_token(self):
...
+        """Get the rules directory path."""
+        return os.path.join(self.nbi_user_dir, 'rules')
+
+    @property
+    def active_rules(self) -> dict:
+        """Get dictionary of active rule states (filename -> bool)."""
+        return self.get('active_rules', {})
+
+    @property
+    def rules_auto_reload(self) -> bool:
```
Comment (mbektas) [inline notebook_intelligence/config.py:140]: not needed at this time

File: notebook_intelligence/config.py [hunk 2]
```diff
@@ -120,3 +120,38 @@ def store_github_access_token(self):
...
+        """Check if rules should be automatically reloaded when changed."""
+        return self.get('rules_auto_reload', True)
+
+    @property
+    def rules_context_injection_enabled(self) -> bool:
```
Comment (mbektas) [inline notebook_intelligence/config.py:145]: remove this

File: notebook_intelligence/models.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
+from dataclasses import dataclass, field
```
Comment (mbektas) [inline notebook_intelligence/models.py:1]: rename this file to ruleset.py

**Review by mbektas — APPROVED** [review]
LGTM! thanks.
